crates/telio-wg/src/adapter/boring.rs

```diff
@@ -126,6 +126,15 @@ impl Adapter for BoringTun {
             fn flush(&mut self) -> io::Result<()> {
                 Ok(())
             }
+
+            // Implement `write_all()` manually because on Apple plafrorms
+            // borngtun writes more bytes than the buffer length returning
+            // number of bytes written > `buf.len()`. This causes panic in the
+            // default `write_all()` implementation.
+            fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+                self.tun.write4(buf);
```

Comment:
Huh, you might be right. Do you suggest subtracting 4 on Apple platforms in `write()`?

Reply:
For me it looks like a bug in the `write4()` function for Apple, it should adjust for its internal headers

Reply:
It looks like borintun itself doesn't care about the number of bytes written
* https://github.com/NordSecurity/boringtun/blob/main/boringtun/src/device/mod.rs#L857
* https://github.com/NordSecurity/boringtun/blob/main/boringtun/src/device/mod.rs#L978

Reply:
Hmm I'm not entirely sure, but my intuition tells if `write_all()` internally just calls `write()` in a loop which we already override here, maybe we could just subtract -4 there? then we wouldn't have to override `write_all()` too?

```
            fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
                Ok(self.tun.write4(buf) - 4)
            }
```

But then comes more fun, what if write4 returns 0 etc.. 😬

Reply:
Looking at it again, since it's only for `inject_reset_packets` and not the main data flow (just the ICMP packets, not streaming youtube), maybe it's ok :)

Reply:
Would doing `sink4.write(ipv4pkgbuf)?;` instead of `sink4.write_all(ipv4pkgbuf)?;` in [firewall::reset_udp_conns()](https://github.com/NordSecurity/libtelio/blob/acdcc7b37be8ee5f625a131b9498b5dbcfce5bbd/crates/telio-firewall/src/firewall.rs#L1200) also work?

Reply:
In that case yes, but the firewall doesn't assume a particular implementation. So it's safer to call `write_all()` from the firewall

Reply:
https://github.com/NordSecurity/boringtun/pull/32